hnn_core/network_builder.py

```diff
@@ -346,109 +365,89 @@ def _gid_assign(self):
+            # with the current node/rank
+            for key in self.p_unique.keys():
                 gid_input = gid + self.net.gid_dict[key][0]
                 _PC.set_gid2node(gid_input, rank)
-                self.net._gid_list.append(gid_input)
+                self._gid_list.append(gid_input)
 
         for gid_base in range(rank, self.net.n_common_feeds, nhosts):
             # shift the gid_base to the common gid
             gid = gid_base + self.net.gid_dict['common'][0]
             # set as usual
             _PC.set_gid2node(gid, rank)
-            self.net._gid_list.append(gid)
+            self._gid_list.append(gid)
         # extremely important to get the gids in the right order
-        self.net._gid_list.sort()
+        self._gid_list.sort()
 
-    def _create_cells_and_feeds(self):
+    def _create_cells_and_feeds(self, threshold):
         """Parallel create cells AND external inputs (feeds)
 
+        NB: _Cell.__init__ calls h.Section -> non-picklable!
+        NB: _ArtificialCell.__init__ calls h.XXX -> non-picklable!
+
         These feeds are spike SOURCES but cells are also targets.
         External inputs are not targets.
         """
-        params = self.net.params
-        # Re-create external feed param dictionaries
-        # Note that the only thing being updated here are the param['prng_*']
-        # values
-        self.net.p_common, self.net.p_unique = create_pext(params,
-                                                           params['tstop'])
 
-        # loop through gids on this node
-        for gid in self.net._gid_list:
+        type2class = {'L2_pyramidal': L2Pyr, 'L5_pyramidal': L5Pyr,
+                      'L2_basket': L2Basket, 'L5_basket': L5Basket}
+        # loop through ALL gids
+        # have to loop over self._gid_list, since this is what we got
+        # on this rank (MPI)
 
-            src_type, src_pos, is_cell = self.net._get_src_type_and_pos(gid)
+        # XXX need mechanism for Builder to keep track of which trial it's on
+        this_trial_event_times =\
+            self.net._tmp_feed_event_times[self.trial_idx]
 
-            # check existence of gid with Neuron
-            if not _PC.gid_exists(gid):
-                msg = ('Source of type %s with ID %d does not exists in '
-                       'Network' % (src_type, gid))
-                raise RuntimeError(msg)
+        for gid in self._gid_list:
+            src_type, src_pos, is_cell = self.net._get_src_type_and_pos(gid)
 
             if is_cell:  # not a feed
                 # figure out which cell type is assoc with the gid
                 # create cells based on loc property
-                # creates a NetCon object internally to Neuron
-                type2class = {'L2_pyramidal': L2Pyr, 'L5_pyramidal': L5Pyr,
-                              'L2_basket': L2Basket, 'L5_basket': L5Basket}
-                Cell = type2class[src_type]
+                # XXX the cell subclasses could use cleaning up!
```

Comment:
XXX is reserved for more specific comments. Usually, we go over them before a release and see if some of them can be knocked off easily.

Reply:
OK, I'll go through the ones I've left and decide either to do something about or remove.